basic setup for pump

maxvar opened the pull request:
basic pump for water and lava, consumes energy to pump
replaces water/lava blocks with cobblestone
range and depth changed in gui, can be sped up and muffled with upgrades
internal tank only for 16 buckets, use attached tanks for major fluid collection

maxvar committed:
basic setup for pump

Ayutac commented on src/main/java/techreborn/blockentity/machine/misc/PumpBlockEntity.java:
```diff
@@ -0,0 +1,211 @@
+				if (blockState.getBlock() == Blocks.WATER) {
+					fluid = Fluids.WATER;
+					soundEvent = SoundEvents.ITEM_BUCKET_EMPTY;
+				} else if (blockState.getBlock() == Blocks.LAVA) {
```
I would make the first check for lava and then the default water, since theoretically we could also pump other fluids. For example we have generating oil lakes.

maxvar replied:
I can swap the checks, but the pump is currently limited to water and lava anyway, I tried to pump oil from an artificially created lake and it would not work, needs some other checks for FlowableFluid or another proper base class for TR liquids.

maxvar replied:
Reworked the logic. Anyway, lava differed from other liquids only in the sound when pumped,

Ayutac commented on src/main/resources/data/techreborn/recipes/crafting_table/machine/pump.json:
```diff
@@ -0,0 +1,25 @@
+	"type": "minecraft:crafting_shaped",
+	"pattern": [
+		"PPP",
+		"PBP",
```
A pump should be a Tier 1 machine?

maxvar replied:
done

Ayutac reviewed (commented):
Haven't tested ingame yet, but the code looks good. Haven't read into the core functionality too deep.

maxvar committed:
added proper outward scanner for blocks

maxvar committed:
added changes from Spotless

maxvar committed:
moved pump to tier2 package
increased energy storage
changed order of checks for lava/water

maxvar committed:
added advanced_machine_frame to the recipe as is proper for tier 2

maxvar committed:
decreased energy consumption
added config to iterate inwards by default

maxvar committed:
fixed one-over block
removed extensive debug logging

maxvar committed:
added display of depth and range to GUI
changed minimum range to 0 (pumps directly below)
extracted fluid detection and sound event logic into separate methods
removed delay in search for the next block to pump
added destruction of fully waterlogged blocks (kelp, stairs, fences etc), they drop their default item

maxvar commented:
can be tested ingame now. all the functions are there.
TODO: replace placeholder textures for the block model (wip)

maxvar committed:
found incorrect reset when range/depth were changed. fixed
optimized check for pumping elapsed time
muffle upgrade now properly silences full tank and not-enough-energy cases
simplified checks for fluid (should work for any properly implemented types)
rewrote checks for failure conditions (failed for quad-overclock setup)

maxvar committed:
added "exhausted" label to GUI
added i10n for GUI labels
removed sound sfx for not-enough-energy and tank-not-accepting-fluid failure events

maxvar commented:
tested in game with water, lava and oil and some waterlogged blocks. that's it. 
all done except the textures.

maxvar committed:
removed debug logging
updated textures with animation

maxvar commented:
a question to maintainers - should I squash the commits and make another PR or you'll review as is and do squash+merge yourself?

Ayutac commented:
we squash-merge ourselves :)

Ayutac commented on src/main/java/techreborn/blockentity/machine/tier2/PumpBlockEntity.java:
```diff
@@ -0,0 +1,350 @@
+		return exhausted;
+	}
+
+	public void setExhausted(boolean v) {
```
better variable name

maxvar replied:
fixed

Ayutac commented on src/main/java/techreborn/blockentity/machine/tier2/PumpBlockEntity.java:
```diff
@@ -0,0 +1,350 @@
+		exhausted = v;
+	}
+
+	private boolean exhausted;
```
all fields should be grouped together instead of being strewn thorough the class

maxvar replied:
fixed

Ayutac commented on src/main/java/techreborn/blockentity/machine/tier2/PumpBlockEntity.java:
```diff
@@ -0,0 +1,350 @@
+		}
+	}
+
+	private int depth;
```
see above

maxvar replied:
fixed

Ayutac commented on src/main/java/techreborn/blockentity/machine/tier2/PumpBlockEntity.java:
```diff
@@ -0,0 +1,350 @@
+		pumpedTargetBlockPos = null;
+	}
+
+	public PumpBlockEntity(BlockPos pos, BlockState state) {
```
usually constructor is first method in a class

maxvar replied:
fixed

Ayutac commented on src/main/java/techreborn/blockentity/machine/tier2/PumpBlockEntity.java:
```diff
@@ -0,0 +1,350 @@
+				//no longer fluid there?
+				if (fluid == Fluids.EMPTY) {
+					//play oops
+					world.playSound(null, this.pos, SoundEvents.BLOCK_DISPENSER_FAIL, SoundCategory.BLOCKS, isMuffled() ? 0.1f : 1.0f, 1.0f);
```
the sound should only be played if the machine is not muffled. Muffle upgrade reduces sound to 0

maxvar replied:
fixed

Ayutac commented on src/main/java/techreborn/blockentity/machine/tier2/PumpBlockEntity.java:
```diff
@@ -0,0 +1,350 @@
+					tank.getFluidInstance().addAmount(FluidValue.BUCKET);
+				}
+				//play sound
+				world.playSound(null, this.pos, SoundEvents.ITEM_BUCKET_EMPTY, SoundCategory.BLOCKS, isMuffled() ? 0.1f : 1.0f, 1.0f);
```
same here about muffler

maxvar replied:
fixed

maxvar committed:
updated textures and animation

maxvar committed:
grouped fields together and moved constructors and classes
rewrote sound selection + muffled pump now gives no sound at all

maxvar committed:
changed replacement blocks for end and nether to endstone and blackstone
added pump block into the functional tab of creative menu

maxvar committed:
added pump block into the functional tab of creative menu

maxvar committed:
conflict resolved

maxvar committed:
Merge branch '1.19' into pump

modmuss50 reviewed (approved):
Looks great! Nice work 👍
